internal/controller/nats/provisioner.go

```diff
@@ -81,18 +88,64 @@ func (r *Reconciler) handleNATSState(ctx context.Context, nats *nmapiv1alpha1.NA
 		return kcontrollerruntime.Result{}, r.syncNATSStatusWithErr(ctx, nats, err, log)
 	}
 
-	if isSTSReady {
-		nats.Status.SetStateReady()
-		nats.Status.SetURL(nmctrlurl.Format(nats.Name, nats.Namespace))
-		events.Normal(r.recorder, nats, nmapiv1alpha1.ConditionReasonDeployed, "StatefulSet is ready and NATS is deployed.")
-	} else {
+	if !isSTSReady {
 		nats.Status.SetWaitingStateForStatefulSet()
+		nats.Status.AvailabilityZonesUsed = 0
+		// record metric.
+		r.collector.RecordAvailabilityZonesUsedMetric(nats.Status.AvailabilityZonesUsed)
+		// publish k8s event.
 		events.Normal(r.recorder, nats, nmapiv1alpha1.ConditionReasonDeploying,
 			"NATS is being deployed, waiting for StatefulSet to get ready.")
 		r.logger.Info("Reconciliation successful: waiting for STS to get ready...")
 		return kcontrollerruntime.Result{RequeueAfter: RequeueTimeForStatusCheck * time.Second}, r.syncNATSStatus(ctx, nats, log)
 	}
 
+	// set status to ready.
+	nats.Status.SetStateReady()
+	nats.Status.SetURL(nmctrlurl.Format(nats.Name, nats.Namespace))
+	events.Normal(r.recorder, nats, nmapiv1alpha1.ConditionReasonDeployed, "StatefulSet is ready and NATS is deployed.")
+
+	// sync status for AvailabilityZones.
+	nats.Status.AvailabilityZonesUsed, err = r.kubeClient.GetNumberOfAvailabilityZonesUsedByPods(ctx,
+		nats.GetNamespace(), getNATSPodsMatchLabels())
+
+	// record metric.
+	r.collector.RecordAvailabilityZonesUsedMetric(nats.Status.AvailabilityZonesUsed)
+
+	switch {
+	case err != nil:
+		nats.Status.UpdateConditionAvailabilityZones(kmetav1.ConditionFalse,
+			nmapiv1alpha1.ConditionReasonProcessingError, err.Error())
+		events.Warn(r.recorder, nats, nmapiv1alpha1.ConditionReasonProcessingError,
+			err.Error())
+		// if zone information is missing, then it is not possible to determine if the pods
+		// are deployed in different availability zones. So we will not set the state to warning.
+		if !errors.Is(err, k8s.ErrNodeZoneLabelMissing) {
+			nats.Status.State = nmapiv1alpha1.StateError
+			return kcontrollerruntime.Result{}, errors.Join(err, r.syncNATSStatus(ctx, nats, log))
+		}
+	case nats.Spec.Cluster.Size < nmmgr.MinClusterSize:
+		// if cluster size is less than 3, then it NATS is not in cluster mode.
+		// Therefore, availability zones do not matter.
+		nats.Status.UpdateConditionAvailabilityZones(kmetav1.ConditionFalse,
+			nmapiv1alpha1.ConditionReasonNotConfigured,
+			"NATS is not configured to run in cluster mode (i.e. spec.cluster.size < 3).")
+	case nats.Status.AvailabilityZonesUsed == nats.Spec.Cluster.Size:
```

Comment:
I believe this can be `case nats.Status.AvailabilityZonesUsed >= nats.Spec.Cluster.Size`.

Reply:
But `nats.Status.AvailabilityZonesUsed > nats.Spec.Cluster.Size` is not possible?